cpp/tensorrt_llm/runtime/worldConfig.cpp

```diff
@@ -65,17 +67,66 @@ void initMpi(nvinfer1::ILogger& logger, int threadMode = MPI_THREAD_FUNNELED)
 
 } // namespace
 
+WorldConfig::WorldConfig(SizeType tensorParallelism, SizeType pipelineParallelism, SizeType rank, SizeType gpusPerNode,
+    std::optional<std::vector<SizeType>> const& deviceIds)
+    : mTensorParallelism{tensorParallelism}
+    , mPipelineParallelism{pipelineParallelism}
+    , mRank{rank}
+    , mGpusPerNode{gpusPerNode}
+    , mDeviceIds{deviceIds.value_or(std::vector<SizeType>(mGpusPerNode))}
+{
+    auto const numDevices = mDeviceIds.size();
+    TLLM_CHECK(numDevices > 0);
+
+    if (!deviceIds.has_value())
+    {
+        mDeviceIds.resize(mGpusPerNode);
+        std::iota(mDeviceIds.begin(), mDeviceIds.end(), 0);
+    }
+    else
+    {
+        // total number is at most mGpusPerNode
+        TLLM_CHECK_WITH_INFO(static_cast<SizeType>(numDevices) <= mGpusPerNode,
+            "Number of device IDs %zu is greater than GPUs per node %d", numDevices, mGpusPerNode);
+
+        // all deviceIds is within the range
+        TLLM_CHECK(*std::max_element(mDeviceIds.begin(), mDeviceIds.end()) < mGpusPerNode);
+        TLLM_CHECK(*std::min_element(mDeviceIds.begin(), mDeviceIds.end()) >= 0);
+
+        // all ids are unique
+        std::set<SizeType> const deviceIdSet(mDeviceIds.begin(), mDeviceIds.end());
+        TLLM_CHECK_WITH_INFO(
+            deviceIdSet.size() == numDevices, "Device IDs are not unique %zu != %zu", deviceIdSet.size(), numDevices);
+
+        // log a warning if device ids are not contiguous
+        if (std::adjacent_find(deviceIdSet.begin(), deviceIdSet.end(), [](auto x, auto y) { return y - x != 1; })
+            != deviceIdSet.end())
+        {
+            TLLM_LOG_WARNING("The user specified device IDs are not contiguous!");
+        }
+        TLLM_LOG_INFO("Using user-specified devices: %s", tc::arr2str(mDeviceIds.data(), numDevices).c_str());
+    }
+
+    TLLM_CHECK(mTensorParallelism > 0);
+    TLLM_CHECK(mPipelineParallelism > 0);
+
+    TLLM_CHECK_WITH_INFO(static_cast<SizeType>(numDevices) >= tensorParallelism * pipelineParallelism,
```

Comment:
There seems to be a mistake here? In a multi-node setup, the product of `mGpusPerNode` and the number of nodes should exceed the total of `TP` multiplied by `PP`. However, in this case, the number of GPUs per node is actually less than the product of `TP` and `PP`
@kaiyux

Reply:
Thanks for the sharp catch. I will discuss with the TensorRT-LLM engineers working on C++ runtime and go back to you later. 

June

Reply:
Hello, is there any progress on this?

Reply:
Hi @leavelet , we did not fully test multi-node support in TensorRT-LLM, it might be working, but there is no guarantee. If you need to try that, we suggest that you remove the check locally so that it won't block you. Thanks very much for your interest on our work!